src/chrome/internal/breakpoints/features/bpAtNotLoadedScriptViaHeuristicSetter.ts

```diff
@@ -0,0 +1,130 @@
+    private readonly _bprToHeuristicBPR = new ValidatedMap<BPRecipeInSource, BPRecipeInUrlRegexp>();
+
+    public constructor(
+        @inject(TYPES.BaseSourceMapTransformer) private readonly _sourceMapTransformer: BaseSourceMapTransformer,
+        @inject(TYPES.IDebuggeeBreakpointsSetter) private readonly _targetBreakpoints: IDebuggeeBreakpointsSetter) { }
+
+    public async addBPRecipe(requestedBP: BPRecipeInSource, eventsConsumer: IEventsConsumer): Promise<void> {
+        const location = await this.getBPRInUrlRegexpPosition(requestedBP);
+
+        // The runtimeLocation is only used in ExistingBPsForJustParsedScriptSetter which is not needed for the heuristic, so we pass an unavailable loaded source
+        const runtimeLocation = new LocationInLoadedSource(new NoLoadedSourceAvailable(), Position.origin);
+
+        const heuristicBPRecipe = mapToUrlRegexp(requestedBP, location.url.textRepresentation, location.position, runtimeLocation);
+
+        this._bprToHeuristicBPR.set(requestedBP, heuristicBPRecipe);
+        await this._targetBreakpoints.setBreakpointByUrlRegexp(heuristicBPRecipe, eventsConsumer);
+    }
+
+    public async removeBPRecipeIfNeeded(requestedBP: BPRecipeInSource): Promise<void> {
+        const heuristicBPR = this._bprToHeuristicBPR.tryGetting(requestedBP);
+        if (heuristicBPR !== undefined) {
+            await this._targetBreakpoints.removeBreakpoint(heuristicBPR);
+            this._bprToHeuristicBPR.delete(requestedBP);
+        }
+    }
+
+    private async getBPRInUrlRegexpPosition(requestedBP: BPRecipeInSource<IBPActionWhenHit>): Promise<LocationInUrl> {
+        const sourceIdentifier = requestedBP.location.resource.sourceIdentifier;
+        const sourceMap = await this._sourceMapTransformer.getSourceMapFromAuthoredPath(sourceIdentifier);
+        if (sourceMap !== null) {
+            const script = new SourceWithSourceMap(sourceMap);
+            const sourceMapper = new MappedSourcesMapper(script, sourceMap);
+            const mappedLocation = sourceMapper.getPositionInScript(requestedBP.location);
+            return new LocationInUrl(<IResourceIdentifier<CDTPScriptUrl>>sourceMap.generatedPath, mappedLocation.enclosingRange.range.start);
+        }
+
+        // We don't know if this cast is correct or not. If it's not, the breakpoint will not bind, as designed
+        return new LocationInUrl(<IResourceIdentifier<CDTPScriptUrl>>sourceIdentifier, requestedBP.location.position);
+    }
+}
+
+/**
+ * We use this file when we have a typescript file, for which we found a source-map using the EagerSourceMapTransformer before the correspondant script
+ * was loaded by CDTP, so we need to use the source-mapping logic, yet we don't have an instance of IScript to do it. We use this instance instead,
+ * that will correctly help map typescript files to javascript files for non .html files.
+ */
+export class SourceWithSourceMap implements IHasSourceMappingInformation {
```

Comment:
Similar thing with this class. Is it really necessary to create a spoof instance of this class just to tell us to start at 0,0?

Reply:
How would you do this differently?